dash-pipeline/utils/dash_pipeline_utils.py

```diff
@@ -0,0 +1,361 @@
+            entry.priority = priority
+
+        req = p4runtime_pb2.ReadRequest()
+        req.device_id = 0
+        entity = req.entities.add()
+        entity.table_entry.CopyFrom(entry)
+        for response in self.stub.Read(req):
+            for entity in response.entities:
+                yield entity.table_entry
+
+    def write(self, entry, update_type):
+        req = p4runtime_pb2.WriteRequest()
+        req.device_id = 0
+        update = req.updates.add()
+        update.type = update_type
+        update.entity.table_entry.CopyFrom(entry)
+        self.stub.Write(req)
+
+
+class P4InternalConfigTable(P4Table):
+    def __init__(self, target=None):
+        super(P4InternalConfigTable, self).__init__(target)
+        self.p4info_table_internal_config = self.p4info.get_table("dash_ingress.dash_lookup_stage.pre_pipeline_stage.internal_config")
+
+    def set(self,
+            neighbor_mac :bytes = None,
+            mac :bytes = None,
+            cpu_mac :bytes = None,
+            flow_enabled :bytes = None):
+        '''
+        Set dash pipeline internal config by updating table entry of internal_config.
+
+        if one argument is not specifed, the action param is not changed in the
+        existing table entry, otherwise set default value in new table entry.
+        '''
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 1
+        match.ternary.value = b'\x00'
+        match.ternary.mask = b'\xff'
+
+        for entry in self.read(self.p4info_table_internal_config.preamble.id, [match], priority = 1):
+            changed = 0
+
+            param = entry.action.action.params[0]
+            if neighbor_mac and neighbor_mac != param.value:
+                param.value = neighbor_mac
+                changed += 1
+
+            param = entry.action.action.params[1]
+            if mac and mac != param.value:
+                param.value = mac
+                changed += 1
+
+            param = entry.action.action.params[2]
+            if cpu_mac and cpu_mac != param.value:
+                param.value = cpu_mac
+                changed += 1
+
+            param = entry.action.action.params[3]
+            if flow_enabled and flow_enabled != param.value:
+                param.value = flow_enabled
+                changed += 1
+
+            if not changed:
+                return # none of change
+
+            self.write(entry, p4runtime_pb2.Update.MODIFY)
+            return
+
+        # Add one entry
+        entry = p4runtime_pb2.TableEntry()
+        entry.table_id = self.p4info_table_internal_config.preamble.id
+        entry.priority = 1
+        entry.match.append(match)
+
+        action_set_internal_config = self.p4info.get_action("dash_ingress.dash_lookup_stage.pre_pipeline_stage.set_internal_config")
+        entry.action.action.action_id = action_set_internal_config.preamble.id
+        action = entry.action.action
+
+        param = action.params.add()
+        param.param_id = 1
```

Comment:
might be better to change this to constant instead of directly using literal/magic number

Reply:
Will do

Reply:
Fixed. Refactor dash_pipeline_utils.py to get param_id from p4info of table action.